### PR TITLE
CMP+CRMF: Add data structures for CMPv3 and OSSL_CRMF_ATTRIBUTETYPEANDVALUE

### PR DESCRIPTION
**Pull request by DDvO**
This adds datatypes and related getter/setter functions supporting the upcoming CMP-Updates RFC (see https://datatracker.ietf.org/doc/html/draft-ietf-lamps-cmp-updates), which defines extensions to CMPv2 and the new CMP version 3.

This PR is a little bulky (including the new docs) but its contents are pretty shallow and schematic.

- [x] documentation is added or updated

### CODE REVIEW

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:14]: Would it be better to move them to a new file include/crypto/crmf.h ?

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:156]: Not sure we need to repeat the comment 3 times.

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:347]: Is this necessary here? IMO it is better to not touch the output argument in case of error.

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:377]: Similarly as above.

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:300]: Similarly as elsewhere.

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:255]: Similarly as elsewhere.

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:433]: Similarly as elsewhere.

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:481]: Similarly as elsewhere.

File: crypto/cmp/cmp_asn.c
Comment (t8m) [inline crypto/cmp/cmp_asn.c:516]: Again.

File: crypto/crmf/crmf_lib.c
Comment (t8m) [inline crypto/crmf/crmf_lib.c:554]: I am not sure this right. I'd prefer keeping the indentation.

File: doc/man3/OSSL_CMP_ATAV_set0.pod
Comment (t8m) [inline doc/man3/OSSL_CMP_ATAV_set0.pod:72]: This should document when these functions were added (3.1)

File: doc/man3/OSSL_CMP_ATAV_set0.pod
Comment (t8m) [inline doc/man3/OSSL_CMP_ATAV_set0.pod:76]: Maybe update the copyright?

File: doc/man3/OSSL_CMP_ITAV_set0.pod
Comment (t8m) [inline doc/man3/OSSL_CMP_ITAV_set0.pod:104]: The pointer also might not be touched. IMO better semantics is to not touch outputs on error, or rather keep the output value undefined in case of error to keep a leeway for future changes.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by kaduk — COMMENTED** [review]
It is perhaps worth asking whether this change should wait to be merged (but nt necessarily to be reviewed/approved) until the final RFC is published.  It does appear to implement new ASN.1 types that are defined by the currently-still-a-draft document, but on a cursory look those appear to be contained as internal to libcrypto (e.g., in `cmp_local.h`).  So I'm not sure there's a "clearly obvious" answer one way or the other.

**Comment by DDvO** [comment]
@kaduk good point to discuss the timeline of such PRs.
This and related PRs aim at implementing CMPv3 as described in https://datatracker.ietf.org/doc/html/draft-ietf-lamps-cmp-updates. This IETF draft has already passed LAMPS WG last call and is currently under IANA review.
It is on the agenda of the IESG telechat on June 2nd, and we expect IESG approval soon thereafter.
Already now, there are no more technical changes to be expected.

For me being both co-author and implementer it was very helpful to do large parts of the OpenSSL-based implementation already when we still had the chance to adapt the specification in case the implementation work brings up problems or ambiguities in details of the spec. Since meanwhile the technical specification is complete, I believe now is good time to start submitting the code such that time pressure before the next OpenSSL feature release (v3.1) can be avoided.

A complication is that the PRs contributing to the CMPv3 implementation (most of which are not yet available) are not independent - there are certain dependencies, in particular, most of them build on the data structures being defined here in this PR. Therefore, it would be very helpful for submitting them if this PR gets merged rather soon.

I agree that merging of CMPv3 feature branches can/should wait until the final approval of CMP-Updates has been done, which we expect for June.
Yet it is pretty likely that the release/publication of the document will be delayed for quite some time just because other documents that it references are not yet released.

Therefore in my view it would be nice if we do not have to wait until CMP-Updates is ultimately published.
In particular, for the PR dependency reason given above, I suggest merging this PR (assuming of course that it will be approved by then) as soon as CMP-Updates got the IESG approval.

File: crypto/cmp/cmp_asn.c
Comment (DDvO) [reply crypto/cmp/cmp_asn.c]: Yeah, I agree it's redundant to give those comments each and every time.
I've removed many of those repetitions.

File: doc/man3/OSSL_CMP_ITAV_set0.pod
Comment (DDvO) [reply doc/man3/OSSL_CMP_ITAV_set0.pod]: I had thought it would be beneficial to initialize the values as far as possible even on error as a precaution, but now I've switched to the design principle you mention here,
(of course, both in the code, and in the documentation).

File: doc/man3/OSSL_CMP_ATAV_set0.pod
Comment (DDvO) [reply doc/man3/OSSL_CMP_ATAV_set0.pod]: Oops, this was a copy&paste mistake, fixed.

File: doc/man3/OSSL_CMP_ATAV_set0.pod
Comment (DDvO) [reply doc/man3/OSSL_CMP_ATAV_set0.pod]: Done.

File: crypto/crmf/crmf_lib.c
Comment (DDvO) [reply crypto/crmf/crmf_lib.c:554]: The new indentation is also used elsewhere,
and it is the one done by the Emacs C mode and checked as correct by `check-format.pl`.

File: crypto/cmp/cmp_asn.c
Comment (DDvO) [reply crypto/cmp/cmp_asn.c]: I've removed this kind of initialization here and also in all other occurrences mentioned below.

**Comment by DDvO** [comment]
Thanks @t8m for your comments.
I've meanwhile handled all of them.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Can someone please do the 2nd review?

**Review by paulidale — COMMENTED** [review]
Well outside my comfort zone.  I'd prefer someone more knowledgeable about this to review this.  If nobody takes up the challenge, I'll have a stab next week (time permitting).

**Comment by kaduk** [comment]
> Well outside my comfort zone. I'd prefer someone more knowledgeable about this to review this. If nobody takes up the challenge, I'll have a stab next week (time permitting).

I have some experience with CMP, but am unlikely to have time in the next week (I start a new role for $dayjob on monday).
Skimming, it seems that this is mostly just about the data types, so a willingness to look at the ASN.1 is the main requirement.

**Comment by DDvO** [comment]
Thank you @paulidale and @kaduk for your comments and your willingness to review this.

It's really "just" some typical ASN.1 structures (where I'm not an expert myself, but simply learned from existing models)
and directly related straightforward C datatype declarations with their getter and setter functions.

The specs of the implemented ASN.1 structures are given as comments in `crypto/cmp/cmp_local.h`, 
such that no particular current (CMP+CRMF) domain knowledge should be needed.
I repeat them here for convenience from [Appendix A.1](https://datatracker.ietf.org/doc/html/draft-ietf-lamps-cmp-updates-18#appendix-A.1) of the spec, removing comments in that ASN.1 module that are not important here:
```
-- CertReqTemplateContent, id-regCtrl-algId, id-regCtrl-algId, and
   -- id-regCtrl-rsaKeyLen were added in CMP Updates [thisRFC]

   CertReqTemplateContent ::= SEQUENCE {
      certTemplate           CertTemplate,
      keySpec                Controls OPTIONAL
      -- Controls  ::= SEQUENCE SIZE (1..MAX) OF AttributeTypeAndValue
      -- as specified in CRMF ([RFC4211](https://datatracker.ietf.org/doc/html/rfc4211))
      }

   id-regCtrl-algId OBJECT IDENTIFIER ::= { id-regCtrl 11 }
      AlgIdCtrl ::= AlgorithmIdentifier

   id-regCtrl-rsaKeyLen OBJECT IDENTIFIER ::= { id-regCtrl 12 }
      RsaKeyLenCtrl ::= INTEGER (1..MAX)
      
   -- RootCaKeyUpdateContent, CRLSource, and CRLStatus were added in
   -- CMP Updates [thisRFC]

   RootCaKeyUpdateContent ::= SEQUENCE {
      newWithNew       CMPCertificate,
      newWithOld   [0] CMPCertificate OPTIONAL,
      oldWithNew   [1] CMPCertificate OPTIONAL
      }

   CRLSource ::= CHOICE {
      dpn          [0] DistributionPointName,
      issuer       [1] GeneralNames }

   CRLStatus ::= SEQUENCE {

   InfoTypeAndValue ::= SEQUENCE {
       infoType               OBJECT IDENTIFIER,
       infoValue              ANY DEFINED BY infoType  OPTIONAL
   }
   -- Example InfoTypeAndValue contents include, but are not limited
   -- to, the following (un-comment in this ASN.1 module and use as
   -- appropriate for a given environment):
 

   --   id-it-caCerts          OBJECT IDENTIFIER ::= {id-it 17}
   --      CaCertsValue            ::= SEQUENCE SIZE (1..MAX) OF CMPCertificate

   --   id-it-rootCaKeyUpdate  OBJECT IDENTIFIER ::= {id-it 18}
   --      RootCaKeyUpdateValue    ::= RootCaKeyUpdateContent

   --   id-it-certReqTemplate  OBJECT IDENTIFIER ::= {id-it 19}
   --      CertReqTemplateValue    ::= CertReqTemplateContent

   --   id-it-rootCaCert       OBJECT IDENTIFIER ::= {id-it 20}
   --      RootCaCertValue         ::= CMPCertificate

   --   id-it-certProfile      OBJECT IDENTIFIER ::= {id-it 21}
   --      CertProfileValue        ::= SEQUENCE SIZE (1..MAX) OF UTF8String

   --   id-it-crlStatusList    OBJECT IDENTIFIER ::= {id-it TBD1}
   --   CRLStatusListValue         ::= SEQUENCE SIZE (1..MAX) OF CRLStatus

   --   id-it-crls             OBJECT IDENTIFIER ::= {id-it TBD2}
   --   CRLsValue                  ::= SEQUENCE SIZE (1..MAX) OF CertificateList
```

**Comment by DDvO** [comment]
@paulidale and @kaduk, three weeks ago I provided more detail that is hopefully of help for reviewing this.
Could either of you please do the 2nd review by next week?

**Comment by DDvO** [comment]
> It is perhaps worth asking whether this change should wait to be merged (but nt necessarily to be reviewed/approved) until the final RFC is published. It does appear to implement new ASN.1 types that are defined by the currently-still-a-draft document, but on a cursory look those appear to be contained as internal to libcrypto (e.g., in `cmp_local.h`). So I'm not sure there's a "clearly obvious" answer one way or the other.

Yesterday the draft got its IESG approval, so the technical contents of CMP Updates (defining CMPv3) are officially accepted.

**Comment by DDvO** [comment]
@openssl/committers, can someone please do the 2nd review?
This is blocking further CMPv3 contributions.

**Comment by kaduk** [comment]
> Yesterday the draft got its IESG approval, so the technical contents of CMP Updates (defining CMPv3) are officially accepted.

Historical project practice is to wait for the actual RFC; just IESG Approval is not quite good enough.

Also, I am sorry I have not been able to review much lately -- I've taken a new role within Akamai and getting up to speed with it does not leave much time for OpenSSL work at the moment.

**Comment by paulidale** [comment]
@mattcaswell @levitte?

**Comment by DDvO** [comment]
I learned today that the current BouncyCastle release 1.71 (for Java) already supports the upcoming CMP RFCs,
i.e., the extensions to CMP v2 General Messages and polling and the new version v3 (mainly for central key generation).

**Comment by DDvO** [comment]
Rebased to fix trivial merge conflct.
Ping for 2nd review.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Rebased to fix merge conflicts.

Meanwhile it turned out that some helper functions require some adaptations for usability
and four more such functions are needed:
```
OSSL_CMP_ATAV_new_algId
OSSL_CMP_ATAV_get0_algId
OSSL_CMP_ATAV_new_rsaKeyLen
OSSL_CMP_ATAV_get_rsaKeyLen
```
which also eliminate the need to export `OSSL_CMP_ATAV_dup` .

**Comment by DDvO** [comment]
Carved out the `certProfile` feature to #21281.

**Comment by DDvO** [comment]
> I'm going to split this up topic-wise, which will likely make it better reviewable.

Closing this PR because meanwhile it has been and is being superseded by topic-wise PRs:
* #19231
* #21129
* #23768
* and one more TBD on requesting a certificate request template, as detailed in [RFC 9483 section 4.3.3](https://www.rfc-editor.org/rfc/rfc9483.html#name-get-certificate-request-tem)

**Pull request closed without merging**
